### PR TITLE
feat(snapshots): use semantic equality for SharedMatrix in snapshot tests

### PR DESCRIPTION
**Pull request by connorskees**
## Description

Two equivalent `SharedMatrices` can differ in exact cell content, but still be semantically equal when referenced by all public APIs. We work around this by special-casing equality for matrices in snapshot tests to rely on cell access exclusively through `getCell`.

[ADO 432](https://dev.azure.com/fluidframework/internal/_workitems/edit/432/)

### CODE REVIEW

File: packages/tools/replay-tool/src/replayMessages.ts [hunk 1]
```diff
@@ -777,3 +814,69 @@ export class ReplayTool {
...
+    const dataStoreRuntime = new MockFluidDataStoreRuntime();
+    dataStoreRuntime.local = true;
+    const deltaConnection = new MockEmptyDeltaConnection();
+
+    const objectStorage1 = MockStorage.createFromSummary(
+        convertToSummaryTreeWithStats(d1.value).summary,
+    );
+    const matrix1 = new SharedMatrix(
+        dataStoreRuntime as any,
+        "123",
```
Comment (Abe27342) [inline packages/tools/replay-tool/src/replayMessages.ts:850]: giving more suggestive ids here would be nice--probably just "1" and "2" is fine. Also I typically like defining an inline helper which takes in something like a summary and an id and gives back the DDS you want (should be a bit shorter than the copy paste, and the semantics are easy enough to understand)

File: packages/utils/tool-utils/src/snapshotNormalizer.ts [hunk 2]
```diff
@@ -153,6 +153,46 @@ export function getNormalizedSnapshot(snapshot: ITree, config?: ISnapshotNormali
     };
 }
 
+function normalizeMatrix(value: ITree): ITree {
```
Comment (Abe27342) [inline packages/utils/tool-utils/src/snapshotNormalizer.ts:156]: is this normalization pass necessary now that we're comparing with semantic equality?

**Review by Abe27342 — COMMENTED** [review]
This generally looks reasonable. One question I have is if the changes here impact test runtime much at all (should just be able to compare a few pipeline runs to figure this out). You may want to get review from someone more familiar with the summary format, since I'm definitely not an expert in the area.

**Comment by Abe27342** [comment]
one other thing--fill out description before completing the PR please (just some basic context about how matrices' internal representation may differ on different clients thus the need for this should be fine)

**Comment by connorskees** [comment]
> fill out description before completing the PR please

Thank you for this reminder! I had intended to fill this out before marking the PR ready for review. I will do so after standup.

File: packages/utils/tool-utils/src/snapshotNormalizer.ts (hunk 2 above)
Comment (connorskees) [reply packages/utils/tool-utils/src/snapshotNormalizer.ts:156]: This normalization pass is still necessary to solve issues for snapshot tests running in back compat mode.

**Review by Abe27342 — APPROVED** [review]

**Pull request merged**